solver/src/driver.rs

```diff
@@ -104,91 +105,57 @@ impl Driver {
             .collect()
     }
 
-    pub async fn single_run(&mut self) -> Result<()> {
-        tracing::debug!("starting single run");
-        let liquidity = self.liquidity_collector.get_liquidity().await?;
-
-        let estimated_prices = self.collect_estimated_prices(&liquidity).await;
-        // Filter limit orders for which we don't have price estimates as they cannot be considered for the objective criterion
-        let (liquidity, removed_orders): (Vec<_>, Vec<_>) =
-            liquidity
-                .into_iter()
-                .partition(|liquidity| match liquidity {
-                    Liquidity::Limit(limit_order) => {
-                        [limit_order.sell_token, limit_order.buy_token]
-                            .iter()
-                            .all(|token| estimated_prices.contains_key(token))
-                    }
-                    Liquidity::Amm(_) => true,
-                });
-        if !removed_orders.is_empty() {
-            tracing::debug!(
-                "pruned {} orders: {:?}",
-                removed_orders.len(),
-                removed_orders,
-            );
-        }
-
-        self.metrics.liquidity_fetched(&liquidity);
-
-        let gas_price = self
-            .gas_price_estimator
-            .estimate()
-            .await
-            .context("failed to estimate gas price")?;
-        tracing::debug!("solving with gas price of {}", gas_price);
-
-        let mut settlements: Vec<(String, Settlement)> =
-            join_all(self.solver.iter().map(|solver| {
-                let liquidity = liquidity.clone();
-                let metrics = &self.metrics;
-                async move {
-                    let label = format!("{}", solver);
-                    let start_time = Instant::now();
-                    let settlement = solver.solve(liquidity, gas_price).await;
-                    metrics.settlement_computed(&label, start_time);
-                    (label, settlement)
-                }
-            }))
-            .await
-            .into_iter()
-            .filter_map(|(solver, settlement)| match settlement {
-                Ok(settlement) => settlement.map(|settlement| (solver, settlement)),
+    async fn run_solvers(
+        &self,
+        liquidity: Vec<Liquidity>,
+        gas_price: f64,
+        prices: &HashMap<H160, BigRational>,
+    ) -> Vec<SolverSettlement> {
+        join_all(self.solver.iter().enumerate().map(|(index, solver)| {
+            let liquidity = liquidity.clone();
+            let metrics = &self.metrics;
+            async move {
+                let start_time = Instant::now();
+                let settlement = solver.solve(liquidity, gas_price).await;
+                metrics.settlement_computed(solver.to_string().as_str(), start_time);
+                (index, settlement)
```

Comment:
Why do we pass the index around instead of a reference to the solver itself?

Reply:
If we used references then the SolverSettlement struct would have to have a lifetime and then result of this function would be tied to &'a Self which means you wouldn't be able to call other &mut self methods. Overall it felt easier to use an index.